robot/impl/resource_manager.go

```diff
@@ -529,6 +530,152 @@ func (manager *resourceManager) completeConfig(
+				if !ok || !gNode.NeedsReconfigure() {
+					return
+				}
+				if !(resName.API.IsComponent() || resName.API.IsService()) {
+					return
+				}
+
+				var verb string
+				conf := gNode.Config()
+				if gNode.IsUninitialized() {
+					verb = "configuring"
+					gNode.InitializeLogger(
+						manager.logger, resName.String(), conf.LogConfiguration.Level,
+					)
+				} else {
+					verb = "reconfiguring"
+				}
+				manager.logger.CInfow(ctx, fmt.Sprintf("Now %s resource", verb), "resource", resName)
+
+				// this is done in config validation but partial start rules require us to check again
+				if _, err := conf.Validate("", resName.API.Type.Name); err != nil {
+					gNode.LogAndSetLastError(
+						fmt.Errorf("resource config validation error: %w", err),
+						"resource", conf.ResourceName(),
+						"model", conf.Model)
+					return
+				}
+				if manager.moduleManager.Provides(conf) {
+					if _, err := manager.moduleManager.ValidateConfig(ctxWithTimeout, conf); err != nil {
+						gNode.LogAndSetLastError(
+							fmt.Errorf("modular resource config validation error: %w", err),
+							"resource", conf.ResourceName(),
+							"model", conf.Model)
+						return
+					}
+				}
+
+				switch {
+				case resName.API.IsComponent(), resName.API.IsService():
+					processResource := func() {
+						defer levelWG.Done()
+
+						newRes, newlyBuilt, err := manager.processResource(ctxWithTimeout, conf, gNode, lr)
+						if newlyBuilt || err != nil {
+							if err := manager.markChildrenForUpdate(resName); err != nil {
+								manager.logger.CErrorw(ctx,
+									"failed to mark children of resource for update",
+									"resource", resName,
+									"reason", err)
+							}
+						}
+
+						if err != nil {
+							gNode.LogAndSetLastError(
+								fmt.Errorf("resource build error: %w", err),
+								"resource", conf.ResourceName(),
+								"model", conf.Model)
+							return
+						}
+
+						// if the ctxWithTimeout fails with DeadlineExceeded, then that means that
+						// resource generation is running async, and we don't currently have good
+						// validation around how this might affect the resource graph. So, we avoid
+						// updating the graph to be safe.
+						if errors.Is(ctxWithTimeout.Err(), context.DeadlineExceeded) {
+							manager.logger.CErrorw(
+								ctx, "error building resource", "resource", conf.ResourceName(), "model", conf.Model, "error", ctxWithTimeout.Err())
+						} else {
+							gNode.SwapResource(newRes, conf.Model)
+						}
+					}
+
+					if !forceSync {
+						// TODO(RSDK-6925): support concurrent processing of resources of
+						// APIs with a maximum instance limit. Currently this limit is
+						// validated later in the resource creation flow and assumes that
+						// each resource is created synchronously to have an accurate
+						// creation count.
+						if c, ok := resource.LookupGenericAPIRegistration(resName.API); ok && c.MaxInstance != 0 {
+							forceSync = true
```

Comment:
Behavorial question:

Given Datamanager is the canonical example of this and every robot has a Datamanager, I wanted to make sure I understand the control flow.

Datamanager I think generally depends on a low level components. So in a hypothetical example it's maybe placed at reverse topological level 3 out of 4. So if we're starting up a robot I expect we'll get this kind of behavior:
* Process level 1 (Resources A, B and C are run in separate goroutines)
* Process level 2 (Resources D and E are run in separate goroutines)
* Process level 3 (Resources F, Datamanager and G)
  * go F()
  * Datamanager() (synchronous)
  * G() (synchronous)
* Process level 4 (Resources H and I)
  * H() (synchronous)
  * I() (synchronous)

Is it intended that everything after Datamanager gets serialized?

Reply:
_Go/Threadsafety_ question:

It occurred to me that maybe this `forceSync = true` assignment was intended to be local to this specific resource. And not modify the `forceSync` parameter passed into function. I don't think that's what's happening here. See this [Go playground run](https://go.dev/play/p/ferfVkJ5n-U).

And if this is the intention -- one "MaxInstance" API serializes everything henceforth -- this assignment creates a data race (in a rather contrived way).

Reply:
ugh, great catch by you and big mistake by me. yeah I intended for only this iteration to be sync

Reply:
done https://github.com/viamrobotics/rdk/pull/3707/commits/0d517e2d26ce8866b5af6a95505d5031b46d6e95